source/module_io/write_libxc_r.cpp

```diff
@@ -0,0 +1,285 @@
+		constexpr double grho_threshold = 1E-10;
+
+		xc_func_set_dens_threshold(&func, rho_threshold);
+
+		// sgn for threshold mask
+		const std::vector<double> sgn = XC_Functional_Libxc::cal_sgn(rho_threshold, grho_threshold, func, nspin, nrxx, rho, sigma);
+
+		// call libxc function
+		// attention: order 432 don't break
+		switch( func.info->family )
+		{
+			case XC_FAMILY_LDA:
+			{
+				switch( order )
+				{
+					case 4:		xc_lda_lxc    ( &func, nrxx, rho.data(), v4rho4.data() );
+					case 3:		xc_lda_kxc    ( &func, nrxx, rho.data(), v3rho3.data() );
+					case 2:		xc_lda_fxc    ( &func, nrxx, rho.data(), v2rho2.data() );
+					case 1:		xc_lda_exc_vxc( &func, nrxx, rho.data(), exc.data(), vrho.data() );
+								break;
+					case 0:		xc_lda_exc    ( &func, nrxx, rho.data(), exc.data() );
+								break;
+					default:	throw std::domain_error("order ="+std::to_string(order)
+									+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+								break;
+				}
+				break;
+			}
+			case XC_FAMILY_GGA:
+			case XC_FAMILY_HYB_GGA:
+			{
+				switch( order )
+				{
+					case 4:		xc_gga_lxc    ( &func, nrxx, rho.data(), sigma.data(), v4rho4.data(), v4rho3sigma.data(), v4rho2sigma2.data(), v4rhosigma3.data(), v4sigma4.data() );
+					case 3:		xc_gga_kxc    ( &func, nrxx, rho.data(), sigma.data(), v3rho3.data(), v3rho2sigma.data(), v3rhosigma2.data(), v3sigma3.data() );
+					case 2:		xc_gga_fxc    ( &func, nrxx, rho.data(), sigma.data(), v2rho2.data(), v2rhosigma.data(), v2sigma2.data() );
+					case 1:		xc_gga_exc_vxc( &func, nrxx, rho.data(), sigma.data(), exc.data(), vrho.data(), vsigma.data() );
+								break;
+					case 0:		xc_gga_exc    ( &func, nrxx, rho.data(), sigma.data(), exc.data() );
+								break;
+					default:	throw std::domain_error("order ="+std::to_string(order)
+									+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+								break;
+				}
+				break;
+			}
+			default:
+			{
+				throw std::domain_error("func.info->family ="+std::to_string(func.info->family)
+					+" unfinished in "+std::string(__FILE__)+" line "+std::to_string(__LINE__));
+				break;
+			}
+		} // end switch( func.info->family )
+	} // end for( xc_func_type &func : funcs )
+
+	auto write_data = [&esolver](
+		const std::string data_name,
+		const std::vector<double> &data,
+		const int number_spin)
+	{
+		for(int is=0; is<number_spin; ++is)
+		{
+			std::ofstream ofs;
+			if(GlobalV::MY_RANK==0)
+			{
+				const std::string folder_name = PARAM.globalv.global_out_dir + "/xc_r/";
+				const std::string command0 =  "test -d " + folder_name + " || mkdir " + folder_name;
+				system( command0.c_str() );
+				const std::string file_name = folder_name + data_name+"_"+std::to_string(is);
+				ofs.open(file_name);
+
+				ofs.unsetf(std::ostream::fixed);
+				ofs << std::setprecision(PARAM.inp.out_xc_r[1]);
+				ofs << std::scientific;
+			}
+
+		  #ifdef __MPI
+			ModuleIO::write_cube_core(

```

Comment:
I intentionally delete the header because it contains a lot of irrelevant information, which other functions or other programs may not be concerned about.